src/server/game/Entities/Player/Player.cpp

```diff
@@ -22425,13 +22425,20 @@ bool Player::BuyItemFromVendorSlot(ObjectGuid vendorguid, uint32 vendorslot, uin
     // check current item amount if it limited
     if (crItem->maxcount != 0)
     {
-        if (creature->GetVendorItemCurrentCount(crItem) < pProto->GetBuyCount() * count)
+        if (creature->GetVendorItemCurrentCount(crItem) < count)
         {
             SendBuyError(BUY_ERR_ITEM_ALREADY_SOLD, creature, item, 0);
             return false;
         }
     }
 
+    // Can only buy full stacks
+    if (count % pProto->GetBuyCount())
```

Comment:
It does not work with default UI but you can do this with addons - I use https://mods.curse.com/addons/wow/gnomishvendorshrinker (heavy warning - might not work on 7.2.5 without modifications, source link https://github.com/TekNoLogic/GnomishVendorShrinker/commits/master)